packages/node/src/integrations/anr/index.ts

```diff
@@ -0,0 +1,172 @@
+import { URL } from 'url';
+import { getCurrentHub } from '@sentry/core';
+import type { Contexts, Event, EventHint, Integration } from '@sentry/types';
+import { dynamicRequire, logger } from '@sentry/utils';
+import type { Worker, WorkerOptions } from 'worker_threads';
+import type { NodeClient } from '../../client';
+import { NODE_VERSION } from '../../nodeVersion';
+import type { Options, WorkerStartData } from './common';
+import { base64WorkerScript } from './worker-script';
+
+const DEFAULT_INTERVAL = 50;
+const DEFAULT_HANG_THRESHOLD = 5000;
+
+type WorkerNodeV14 = Worker & { new (filename: string | URL, options?: WorkerOptions): Worker };
+
+type WorkerThreads = {
+  Worker: WorkerNodeV14;
+};
+
+function log(message: string, ...args: unknown[]): void {
+  logger.log(`[ANR] ${message}`, ...args);
+}
+
+/**
+ * We need to use dynamicRequire because worker_threads is not available in node < v12 and webpack error will when
+ * targeting those versions
+ */
+function getWorkerThreads(): WorkerThreads {
+  return dynamicRequire(module, 'worker_threads');
+}
+
+/**
+ * Gets contexts by calling all event processors. This relies on being called after all integrations are setup
+ */
+async function getContexts(client: NodeClient): Promise<Contexts> {
+  let event: Event | null = { message: 'ANR' };
+  const eventHint: EventHint = {};
+
+  for (const processor of client.getEventProcessors()) {
+    if (event === null) break;
+    event = await processor(event, eventHint);
+  }
+
+  return event?.contexts || {};
+}
+
+interface InspectorApi {
+  open: (port: number) => void;
+  url: () => string | undefined;
+}
+
+/**
+ * Starts the node debugger and returns the inspector url.
+ *
+ * When inspector.url() returns undefined, it means the port is already in use so we try the next port.
+ */
+function startInspector(startPort: number = 9229): string | undefined {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const inspector: InspectorApi = require('inspector');
+  let inspectorUrl: string | undefined = undefined;
+  let port = startPort;
+
+  while (inspectorUrl === undefined && port < startPort + 100) {
+    inspector.open(port);
+    inspectorUrl = inspector.url();
+    port++;
+  }
+
+  return inspectorUrl;
+}
+
+/**
+ * Starts a thread to detect App Not Responding (ANR) events
+ */
+export class Anr implements Integration {
+  public name: string = 'Anr';
+
+  public constructor(private readonly _options: Partial<Options> = {}) {}
```

Comment:
the `| undefined` type is assumed from using the default value on the arg.

Reply:
I think this is because we have very loose Typescript configuration. By default `= {}` corresponds to undefined value, but the type doesn't reflect it. I think it should.